backend/test/integration/job_api_test.go

```diff
@@ -97,24 +106,23 @@ func (s *JobApiTestSuite) TestJobApis() {
 	createJobRequest := &jobparams.CreateJobParams{Body: &job_model.APIJob{
 		Name:        "hello world",
 		Description: "this is hello world",
-		PipelineSpec: &job_model.APIPipelineSpec{
-			PipelineID: helloWorldPipeline.ID,
-		},
 		ResourceReferences: []*job_model.APIResourceReference{
 			{Key: &job_model.APIResourceKey{Type: job_model.APIResourceTypeEXPERIMENT, ID: helloWorldExperiment.ID},
 				Relationship: job_model.APIRelationshipOWNER},
+			{Key: &job_model.APIResourceKey{Type: job_model.APIResourceTypePIPELINEVERSION, ID: helloWorldPipelineVersion.ID},
```

Comment:
Discussion:
Shall we keep existing tests and add a new one testing using pipeline version?
So that we can keep test coverage for using pipelines in Job.

Reply:
In fact, we are going  to deprecate the usage of pipeline id to create run/job. And all runs  and jobs must be from some version. The to-be-deprecated approach to creating from pipeline is actually using its default version as well.

Reply:
I see.
I'm curious why we are deprecating that usage, I thought we are going to be backward-compatible.

Reply:
There is a little bit of history. Once upon a time, we are trying to follow the practice of CMLE api.  In that approach, a pipeline is allowed to contain no pipeline versions at all. Then we figure it might be more flexible and more extensible if  we always go from versions.  No matter we allow pipeline to have empty versions or not, creating from versions works. And moreover, we always have a default version, so from the UI, user open a pipeline, its details page actually shows its default version and when run is created from there, we naturally use default version to create run. And blahblah... and also, if we really really really just want pipeline, it should be specified in resource reference field instead of the pipeline spec field, since that resource reference field is meant and added to have a uniform way to specify dependencies between our objects/resources... 

So in short, specifying pipeline in the pipeline spec is going to be deprecated for actually two reasons: (1) we want creation from pipeline versions (2) we want to use resource reference field.

Reply:
I see, so it will be supported to specify pipeline id from resource references? (is it already supported?)
So it's still a breaking change, but we have a migration path, is that right?

Do you have a tracking list for TODO items like updating SDK to not use the spec field?
e.g. here https://github.com/kubeflow/pipelines/blob/119e32910869697e91d749065626c1ff3c66b262/sdk/python/kfp/_client.py#L333-L338

Also, the pipeline spec field also accepts json string of the entire pipeline manifest to run. I think that usage is required for cloning runs. Will we deprecate that?
Looks like you have all UI usages of pipeline_spec with pipeline_id, SGTM.

Reply:
The two reasons are in AND relationship. So we don't add support for pipeline id in resource reference field.

And this is not a breaking change. Since 0.1.38-ish, all runs are referring pipeline versions. If a run is created  before user upgrading to 0.1.38, all runs can still be retrieved correctly, since we keep these old runs' use case when we GetRun. Therefore, we may have runs of these two types in our KFP for a while, but as I said when new runs are getting created, they are all from versions.

Reply:
As for the SDK, all we can do is to notify the users to revise their scripts. E.g., (1) in release note and (2) printing out warning message when they take the approach.

Reply:
Also, an opportunity for us to totally remove all legacies is to introduce breaking changes (that is, surface the internal change to the interface) at KFP 1.0......

Reply:
I agree on these, just want to double check if you have put the following in your plan like:
* adding deprecation notification in sdk methods that we will no longer support (I believe we should do this as early as possible, it's not far away from KFP 1.0's final release.)
* confirm with the team that we will have a breaking change

> So even when we allow the user still uses pipeline id in pipeline spec, internally we'll still record the run with a reference to the default version at the creation time. Hence implement the TODO at

For this part, it's totally reasonable, but it doesn't convince me we need to stop people from using pipeline id to create runs. Your mentioned recording pipeline version id at run creation time sounds like the right fix (and is not hacky at all).

Reply:
Two things: 

(1) the breaking change is optional, we can choose not to. That's my original plan (not to make the sdk client interface change). When we keep the SDK client interface the same, we only have a different implementation underhood (use resource reference instead of pipeline id in spec).  

(2) the reason I talked about breaking change is IF we want it, we can leverage KF 1.0 as a proper time to do.

Reply:
discussed offline: we will fix sdk to workaround the breaking change